File: backend/app/api/v1/okr.py
import re

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.orm import Session

from app.api.v1.deps import csrf_protect, get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.ai import MonthlyReportScoreOut, OkrReviewFullOut
from app.schemas.okr import (
    KeyResultCreate,
    KeyResultProgressCreate,
    KeyResultProgressOut,
    KeyResultUpdate,
    MonthlyReportSectionOut,
    MonthlyReportSectionUpdate,
    ObjectiveCreate,
    ObjectiveOut,
    ObjectiveUpdate,
    OkrOrderUpdate,
    OkrCommentCreate,
    OkrCommentOut,
    OkrCommentUpdate,
    OkrMonthOut,
)
from app.services.ai.provider import AiProviderError, AiProviderNotConfigured
from app.services.ai_serialize import serialize_monthly_report_score, serialize_okr_review
from app.services.ai_service import AiService
from app.services.okr_service import OkrService

router = APIRouter(prefix="/okr", tags=["okr"])

_MONTH_RE = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")


def _valid_month(month: str) -> str:
    if not _MONTH_RE.match(month):
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="month must be YYYY-MM")
    return month


@router.get("", response_model=OkrMonthOut)
def get_month(
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).get_month(_valid_month(month))


@router.post("/objectives", response_model=ObjectiveOut, dependencies=[Depends(csrf_protect)])
def create_objective(
    payload: ObjectiveCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).create_objective(payload)


@router.patch("/objectives/{objective_id}", response_model=ObjectiveOut, dependencies=[Depends(csrf_protect)])
def update_objective(
    objective_id: int,
    payload: ObjectiveUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).update_objective(objective_id, payload)


@router.post("/objectives/reorder", response_model=OkrMonthOut, dependencies=[Depends(csrf_protect)])
def reorder_objectives(
    payload: OkrOrderUpdate,
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).reorder_objectives(_valid_month(month), payload.ids)


@router.delete(
    "/objectives/{objective_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(csrf_protect)],
)
def delete_objective(
    objective_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> Response:
    OkrService(db, user).delete_objective(objective_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/objectives/{objective_id}/key-results",
    response_model=ObjectiveOut,
    dependencies=[Depends(csrf_protect)],
)
def add_key_result(
    objective_id: int,
    payload: KeyResultCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).add_key_result(objective_id, payload)


@router.patch("/key-results/{kr_id}", response_model=ObjectiveOut, dependencies=[Depends(csrf_protect)])
def update_key_result(
    kr_id: int,
    payload: KeyResultUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).update_key_result(kr_id, payload)


@router.post("/objectives/{objective_id}/key-results/reorder", response_model=ObjectiveOut, dependencies=[Depends(csrf_protect)])
def reorder_key_results(
    objective_id: int,
    payload: OkrOrderUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).reorder_key_results(objective_id, payload.ids)


@router.delete("/key-results/{kr_id}", response_model=ObjectiveOut, dependencies=[Depends(csrf_protect)])
def delete_key_result(
    kr_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).delete_key_result(kr_id)


@router.get("/key-results/{kr_id}/progress", response_model=list[KeyResultProgressOut])
def list_key_result_progress(
    kr_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list:
    return OkrService(db, user).list_key_result_progress(kr_id)


@router.post(
    "/key-results/{kr_id}/progress",
    response_model=KeyResultProgressOut,
    dependencies=[Depends(csrf_protect)],
)
def create_key_result_progress(
    kr_id: int,
    payload: KeyResultProgressCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return OkrService(db, user).create_key_result_progress(kr_id, payload)


@router.get("/objectives/{objective_id}/comments", response_model=list[OkrCommentOut])
def list_objective_comments(
    objective_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[dict]:
    return OkrService(db, user).list_objective_comments(objective_id)


@router.post(
    "/objectives/{objective_id}/comments",
    response_model=OkrCommentOut,
    dependencies=[Depends(csrf_protect)],
)
def create_objective_comment(
    objective_id: int,
    payload: OkrCommentCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).create_objective_comment(objective_id, payload)


@router.get("/key-results/{kr_id}/comments", response_model=list[OkrCommentOut])
def list_key_result_comments(
    kr_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[dict]:
    return OkrService(db, user).list_key_result_comments(kr_id)


@router.post(
    "/key-results/{kr_id}/comments",
    response_model=OkrCommentOut,
    dependencies=[Depends(csrf_protect)],
)
def create_key_result_comment(
    kr_id: int,
    payload: OkrCommentCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).create_key_result_comment(kr_id, payload)


@router.patch(
    "/comments/{comment_id}",
    response_model=OkrCommentOut,
    dependencies=[Depends(csrf_protect)],
)
def update_comment(
    comment_id: int,
    payload: OkrCommentUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).update_comment(comment_id, payload)


@router.delete(
    "/comments/{comment_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(csrf_protect)],
)
def delete_comment(
    comment_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> Response:
    OkrService(db, user).delete_comment(comment_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/review", response_model=OkrReviewFullOut)
def get_review(
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> OkrReviewFullOut:
    _valid_month(month)
    return serialize_okr_review(AiService(db, user).get_okr_review(month))


@router.post("/review/generate", response_model=OkrReviewFullOut, dependencies=[Depends(csrf_protect)])
def generate_review(
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> OkrReviewFullOut:
    _valid_month(month)
    service = AiService(db, user)
    if not service.get_flags().okr_review_enabled:
        return OkrReviewFullOut(status="not_enabled")
    try:
        row = service.generate_okr_review(month)
    except AiProviderNotConfigured:
        return OkrReviewFullOut(status="not_enabled")
    except AiProviderError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    return serialize_okr_review(row)


@router.get("/monthly-report/score", response_model=MonthlyReportScoreOut)
def get_monthly_report_score(
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> MonthlyReportScoreOut:
    _valid_month(month)
    return serialize_monthly_report_score(
        AiService(db, user).get_monthly_report_score(month)
    )


@router.post(
    "/monthly-report/score/generate",
    response_model=MonthlyReportScoreOut,
    dependencies=[Depends(csrf_protect)],
)
def generate_monthly_report_score(
    month: str = Query(..., description="YYYY-MM"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> MonthlyReportScoreOut:
    _valid_month(month)
    service = AiService(db, user)
    if not service.get_flags().okr_review_enabled:
        return MonthlyReportScoreOut(status="not_enabled")
    try:
        row = service.generate_monthly_report_score(month)
    except AiProviderNotConfigured:
        return MonthlyReportScoreOut(status="not_enabled")
    except AiProviderError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    return serialize_monthly_report_score(row)


@router.patch(
    "/monthly-report/sections/{section_id}",
    response_model=MonthlyReportSectionOut,
    dependencies=[Depends(csrf_protect)],
)
def update_report_section(
    section_id: int,
    payload: MonthlyReportSectionUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict:
    return OkrService(db, user).update_section(section_id, payload)

File: frontend/src/api/okr.ts
import type {
  KeyResultProgressOut,
  MonthlyReportSectionOut,
  MonthlyReportScoreOut,
  ObjectiveOut,
  OkrMonthOut,
  OkrReviewFullOut,
  OkrCommentOut,
} from '@/types/api';
import { api } from './client';

export function getOkrMonth(month: string): Promise<OkrMonthOut> {
  return api<OkrMonthOut>(`/okr?month=${encodeURIComponent(month)}`);
}

export interface KeyResultInput {
  title: string;
  progress?: number;
}

export interface CreateObjectiveInput {
  month: string;
  title: string;
  key_results?: KeyResultInput[];
}

export function createObjective(input: CreateObjectiveInput): Promise<ObjectiveOut> {
  return api<ObjectiveOut>('/okr/objectives', { method: 'POST', body: input });
}

export interface UpdateObjectiveInput {
  title?: string;
}

export function updateObjective(id: number, input: UpdateObjectiveInput): Promise<ObjectiveOut> {
  return api<ObjectiveOut>(`/okr/objectives/${id}`, { method: 'PATCH', body: input });
}

export function reorderObjectives(month: string, ids: number[]): Promise<OkrMonthOut> {
  return api<OkrMonthOut>(`/okr/objectives/reorder?month=${encodeURIComponent(month)}`, { method: 'POST', body: { ids } });
}

export function deleteObjective(id: number): Promise<null> {
  return api<null>(`/okr/objectives/${id}`, { method: 'DELETE' });
}

export function addKeyResult(objectiveId: number, input: KeyResultInput): Promise<ObjectiveOut> {
  return api<ObjectiveOut>(`/okr/objectives/${objectiveId}/key-results`, {
    method: 'POST',
    body: input,
  });
}

export interface UpdateKeyResultInput {
  title?: string;
  progress?: number;
}

export function updateKeyResult(krId: number, input: UpdateKeyResultInput): Promise<ObjectiveOut> {
  return api<ObjectiveOut>(`/okr/key-results/${krId}`, { method: 'PATCH', body: input });
}

export function reorderKeyResults(objectiveId: number, ids: number[]): Promise<ObjectiveOut> {
  return api<ObjectiveOut>(`/okr/objectives/${objectiveId}/key-results/reorder`, { method: 'POST', body: { ids } });
}

export function deleteKeyResult(krId: number): Promise<ObjectiveOut> {
  return api<ObjectiveOut>(`/okr/key-results/${krId}`, { method: 'DELETE' });
}

export interface CreateKeyResultProgressInput {
  note: string;
  progress_date: string;
}

export function listKeyResultProgress(krId: number): Promise<KeyResultProgressOut[]> {
  return api<KeyResultProgressOut[]>(`/okr/key-results/${krId}/progress`);
}

export function createKeyResultProgress(
  krId: number,
  input: CreateKeyResultProgressInput,
): Promise<KeyResultProgressOut> {
  return api<KeyResultProgressOut>(`/okr/key-results/${krId}/progress`, {
    method: 'POST',
    body: input,
  });
}

export type OkrCommentTarget = 'objective' | 'key-result';

function commentPath(target: OkrCommentTarget, targetId: number): string {
  return target === 'objective'
    ? `/okr/objectives/${targetId}/comments`
    : `/okr/key-results/${targetId}/comments`;
}

export function listOkrComments(
  target: OkrCommentTarget,
  targetId: number,
): Promise<OkrCommentOut[]> {
  return api<OkrCommentOut[]>(commentPath(target, targetId));
}

export function createOkrComment(
  target: OkrCommentTarget,
  targetId: number,
  content: string,
): Promise<OkrCommentOut> {
  return api<OkrCommentOut>(commentPath(target, targetId), {
    method: 'POST',
    body: { content },
  });
}

export function updateOkrComment(commentId: number, content: string): Promise<OkrCommentOut> {
  return api<OkrCommentOut>(`/okr/comments/${commentId}`, {
    method: 'PATCH',
    body: { content },
  });
}

export function deleteOkrComment(commentId: number): Promise<null> {
  return api<null>(`/okr/comments/${commentId}`, { method: 'DELETE' });
}

export function getReview(month: string): Promise<OkrReviewFullOut> {
  return api<OkrReviewFullOut>(`/okr/review?month=${encodeURIComponent(month)}`);
}

export function generateReview(month: string): Promise<OkrReviewFullOut> {
  return api<OkrReviewFullOut>(`/okr/review/generate?month=${encodeURIComponent(month)}`, {
    method: 'POST',
  });
}

export function getMonthlyReportScore(month: string): Promise<MonthlyReportScoreOut> {
  return api<MonthlyReportScoreOut>(
    `/okr/monthly-report/score?month=${encodeURIComponent(month)}`,
  );
}

export function generateMonthlyReportScore(month: string): Promise<MonthlyReportScoreOut> {
  return api<MonthlyReportScoreOut>(
    `/okr/monthly-report/score/generate?month=${encodeURIComponent(month)}`,
    { method: 'POST' },
  );
}

export interface UpdateSectionInput {
  title?: string;
  content_html?: string | null;
  content_json?: string | null;
}

export function updateReportSection(
  sectionId: number,
  input: UpdateSectionInput,
): Promise<MonthlyReportSectionOut> {
  return api<MonthlyReportSectionOut>(`/okr/monthly-report/sections/${sectionId}`, {
    method: 'PATCH',
    body: input,
  });
}
